src/lib.rs

```diff
@@ -482,10 +500,61 @@ impl Octocrab {
         self.execute(request).await
     }
 
+    #[cfg(not(feature = "rate-limiting"))]
     /// Execute the given `request` octocrab's Client.
     pub async fn execute(&self, request: reqwest::RequestBuilder) -> Result<reqwest::Response> {
         request.send().await.context(error::Http)
     }
+
+    #[cfg(feature = "rate-limiting")]
+    /// Execute the given `request` octocrab's Client.
+    pub async fn execute(&self, request: reqwest::RequestBuilder) -> Result<reqwest::Response> {
+        if let Some(delay) = {
+            let mut limiter = self.rate_limiter
+                .as_ref()
+                .lock()
+                .expect("Poisoned rate limiter mutex");
+            limiter.request_delay()
+        } {
+            delay.await;
+        }
+
+        {
+            self.rate_limiter
+                .as_ref()
+                .lock()
+                .expect("Poisoned rate limiter mutex")
+                .register_request();
+        }
+
+        let request_cloned = request.try_clone().unwrap(); // we never use streams in bodies
```

Comment:
Instead of trying to send the request and seeing if it fails, I would instead call the `/rate_limit` endpoint and see how much is left. If you'd also like the public method for this, I would put it under `MetaHandler` struct similar how the rest the API is laid out. https://developer.github.com/v3/rate_limit/

Reply:
This was actually meant as a fallback, in case another source of requests hits the rate limit. In other cases, the `limiter.request_delay().await` should wait if we are rate-limited.